### ISSUE TITLE
Confusion about gt_bboxes_ignore

### ISSUE THREAD

**Issue by chunchet-ng** [open-mmlab/mmdetection#5668]
Hi there,

I have a few questions about the confusing behavior of "gt_bboxes_ignore" parameter in [fcos_head.py](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/dense_heads/fcos_head.py).

1. On [line 170 of fcos_head.py](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/models/dense_heads/fcos_head.py#L170), I noticed that the "gt_bboxes_ignore" is not being used under the loss function. However, by referring to [line 217 of anchor_head.py](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/models/dense_heads/anchor_head.py#L217), I found out that the actual function processing this param is the [MaxIoUAssigner](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/core/bbox/assigners/max_iou_assigner.py#L60), according to this [FCOS config](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/configs/fcos/fcos_r50_caffe_fpn_gn-head_1x_coco.py#L47). The strange part is that I neither find any relevant info about using this assigner in fcos_head.py nor finding "gt_bboxes_ignore" being used anywhere in the code, hence this might lead to the confusion of whether the "gt_bboxes_ignore" is being utilized?
[Similar question](https://github.com/open-mmlab/mmdetection/pull/2935#issuecomment-643102252) has been asked without any reply, it seems apparent on the surface that the param is not supported. However, a [quick test](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_dense_heads/test_fcos_head.py) shows otherwise, setting "gt_bboxes_ignore" the same as "gt_bboxes" will lead to the same result as if ground truth is empty, meaning that the ground truth is ignored.
2. If the respective ground truth is to be ignored (loss is not being calculated for this sample, not positive and not negative), shouldn't it return 0 for all losses? According to [this comment](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/tests/test_models/test_dense_heads/test_fcos_head.py#L44), why the cls loss should be nonzero?
3. What is the exact purpose of using [match_low_quality](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/core/bbox/assigners/max_iou_assigner.py#L184)? Doesn't it clash with the purpose of "gt_bboxes_ignore"? For instance, there are two predictions against one ignored ground truth, the assigned ground truth indexes will be (-1,\<class of gt\>) and this behavior is incorrect.
4. Are samples being assigned as -1 [negative samples](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/core/bbox/assigners/max_iou_assigner.py#L63) or as ignored samples? Where [this line](https://github.com/open-mmlab/mmdetection/blob/9921ee33bd93dab5ec21e3633927e9ccd4f35a80/mmdet/core/bbox/assigners/max_iou_assigner.py#L69) specifies that index 0 is representing negative/background samples.
5. Where is the actual implementation of samples with class -1 that will return 0 for bbox and centerness loss? What I understand is that the assigner will assign predictions to ground truth, but I cannot find further processing of "gt_bboxes_ignore" during loss calculation.

My main purpose is to train a model by specifying which bbox should be ignored and not considered for loss calculation, where I faced this issue while doing so. Any help is greatly appreciated.

Thank you in advance.

**Comment by chunchet-ng** [comment]
Hi @jshilong, can you please help me out on this issue?

**Comment by chunchet-ng** [comment]
Hi @hellock, can you please help me out on this issue?

**Comment by RangiLyu** [comment]
`gt_bboxes_ignore` is not used in FCOS. Some target assigners use `gt_bboxes_ignore` to ignore some box annotations, but FCOS does not use these assigners. I know it is confusing that there is a MaxIoUAssigner in the config of FCOS but not being used. This is a legacy issue. At that time, if not setting an assigner in the config, the code would raise an error. So the MaxIoUAssigner in FCOS config is just a placeholder to avoid raising an error.

**Comment by chunchet-ng** [comment]
Thank you for your clarification on question 1. 

What about the remaining questions 2, 3, 4, and 5? Could you please help to explain them as well?
